### PR TITLE
add support for comparison and text database functions

### PR DESCRIPTION
**Pull request by timgraham**

### CODE REVIEW

**Review by Jibola — APPROVED** [review]

**Review by Jibola — CHANGES_REQUESTED** [review]
Two points and then everything else looks good!

File: django_mongodb/features.py [hunk 1]
```diff
@@ -50,6 +56,12 @@ class DatabaseFeatures(BaseDatabaseFeatures):
         # the result back to UTC.
         "db_functions.datetime.test_extract_trunc.DateFunctionWithTimeZoneTests.test_trunc_func_with_timezone",
         "db_functions.datetime.test_extract_trunc.DateFunctionWithTimeZoneTests.test_trunc_timezone_applied_before_truncation",
+        # Coalesce() with expressions doesn't generate correct query.
+        "db_functions.comparison.test_coalesce.CoalesceTests.test_mixed_values",
+        # $strLenCP requires a string argument, found: null
+        "db_functions.text.test_length.LengthTests.test_basic",
+        # $and must be an array
+        "db_functions.tests.FunctionTests.test_function_as_filter",
```
[the remark this answers is not in the document]
Comment (timgraham) [reply django_mongodb/features.py]: Yes, if feasible. Actulaly, in the latest push I added a workaround for $strLenCP crashing on null:
`{"$cond": {"if": {"$eq": [lhs_mql, None]}, "then": None, "else": {"$strLenCP": lhs_mql}}}`.

File: django_mongodb/functions.py [hunk 2]
```diff
@@ -62,19 +105,58 @@ def func(self, compiler, connection):
...
     # This function is usually log(base, num) but on MongoDB it's log(num, base).
     clone = self.copy()
     clone.set_source_expressions(self.get_source_expressions()[::-1])
     return func(clone, compiler, connection)
 
 
+def null_if(self, compiler, connection):
+    """Return expr1 if expr1==expr2 else None."""
+    expr1, expr2 = (expr.as_mql(compiler, connection) for expr in self.get_source_expressions())
+    return {"$cond": {"if": {"$eq": [expr1, expr2]}, "then": None, "else": expr1}}
```
[the remark this answers is not in the document]
Comment (timgraham) [reply django_mongodb/functions.py]: Well spotted

File: django_mongodb/functions.py [hunk 3]
```diff
@@ -62,19 +105,58 @@ def func(self, compiler, connection):
...
+
+
+def substr(self, compiler, connection):
+    lhs = process_lhs(self, compiler, connection)
+    # The starting index is zero-indexed on MongoDB rather than one-indexed.
+    lhs[1] = {"$add": [lhs[1], -1]}
+    # If no limit is specified, use a large value since $substrCP requires one.
+    if len(lhs) == 2:
+        lhs.append(99999999)
+    return {"$substrCP": lhs}
```
[the remark this answers is not in the document]
Comment (timgraham) [reply django_mongodb/functions.py]: Okay

**Review by WaVEV — APPROVED** [review]

**Review by Jibola — APPROVED** [review]

**Pull request merged**